cloudinit/osys/windows/network.py

```diff
@@ -0,0 +1,415 @@
+# Copyright 2015 Cloudbase Solutions Srl
+# Copyright 2015 Canonical Ltd.
+# License TBD
+
+import ctypes
+from ctypes import wintypes
+import logging
+import os
+import subprocess
+
+import six
+from six.moves import winreg
+from win32com import client
+import wmi
+
+from cloudinit.osys import network
+from cloudinit.osys.windows import general
+from cloudinit.osys.windows.util import iphlpapi
+from cloudinit.osys.windows.util import kernel32
+from cloudinit.osys.windows.util import ws2_32
+
+
+_MIB_IPPROTO_NETMGMT = 3
+_FW_IP_PROTOCOL_TCP = 6
+_FW_IP_PROTOCOL_UDP = 17
+_FW_SCOPE_ALL = 0
+_PROTOCOL_TCP = "TCP"
+_PROTOCOL_UDP = "UDP"
+_ComputerNamePhysicalDnsHostname = 5
+LOG = logging.getLogger(__file__)
+
+
+def _format_mac_address(phys_address, phys_address_len):
+    mac_address = ""
+    for i in range(0, phys_address_len):
+        b = phys_address[i]
+        if mac_address:
+            mac_address += ":"
+        mac_address += "%02X" % b
+    return mac_address
+
+
+def _socket_addr_to_str(socket_addr):
+    addr_str_len = wintypes.DWORD(256)
+    addr_str = ctypes.create_unicode_buffer(256)
+
+    ret_val = ws2_32.WSAAddressToStringW(
+        socket_addr.lpSockaddr,
+        socket_addr.iSockaddrLength,
+        None, addr_str, ctypes.byref(addr_str_len))
+    if ret_val:
+        raise Exception("WSAAddressToStringW failed: %s"
+                        % ws2_32.WSAGetLastError())
+
+    return addr_str.value
+
+
+def _get_registry_dhcp_server(adapter_name):
+    with winreg.OpenKey(
+            winreg.HKEY_LOCAL_MACHINE,
+            "SYSTEM\\CurrentControlSet\\Services\\" +
+            "Tcpip\\Parameters\\Interfaces\\%s" % adapter_name, 0,
+            winreg.KEY_READ) as key:
+        try:
+            dhcp_server = winreg.QueryValueEx(key, "DhcpServer")[0]
+            if dhcp_server == "255.255.255.255":
```

Comment:
Most likely. We don't have support for this in cloudbase-init, that's why it's not present at the moment.